api/src/main/java/org/eclipse/microprofile/reactive/messaging/Message.java

```diff
@@ -80,4 +80,29 @@ public T getPayload() {
     default CompletionStage<Void> ack() {
         return CompletableFuture.completedFuture(null);
     }
+
+    /**
+     * Returns an object of the specified type to allow access to the connector-specific {@link Message} implementation.
+     * If the {@link Message} implementation does not support the target class, an {@link IllegalArgumentException}
+     * should be raised.
+     * The default implementation only supports the {@link Message} class as argument. When a connector provides
+     * its own {@link Message} implementation, it should override this method to support the specific type.
+     *
+     * @param unwrapType the class of the object to be returned, must not be {@code null}
+     * @param <C> the target type
+     * @return an instance of the specified class
+     * @throws IllegalArgumentException if the current {@link Message} instance does not support the call
+     */
+    @SuppressWarnings({"unchecked"})
+    default <C> C unwrap(Class<C> unwrapType) {
+        if (unwrapType == null) {
+            throw new NullPointerException("The target class must not be `null`");
+        }
+        if (Message.class.equals(unwrapType)) {
+            return (C) this;
+        }
+        throw new IllegalArgumentException("Cannot unwrap an instance of " + this.getClass().getName()
+            + " to " + unwrapType.getName());
```

Comment:
Could also consider just calling `Class.cast()` catching the cast exception and creating your `IllegalArgumentException`?

Reply:
implementations may not use a cast, but here we can.

Reply:
If we use:

```
try {
            return unwrapType.cast(this);
        }
        catch (ClassCastException e) {
            throw new IllegalArgumentException("Cannot unwrap an instance of " + this.getClass().getName()
                + " to " + unwrapType.getName(), e);
        }
```

No more issue with the `.equals`

Reply:
Ah yes, ok, I read the above line `return (C) this` and thought we were casting ourselves to `C`, but in fact you're only doing that if `C` is actually `Message`

Reply:
I was so slow understanding the code that by the time I finished writing it, you'd changed it to how I thought it worked in the first place.

If you change it like this, does the javadoc need updating to say that the default implementation will try to cast the object to the requested type?

Reply:
I've updated the javadoc.